### PR TITLE
Feature/v2/windows migration adding windows code

### PR DESCRIPTION
**Pull request by klugorosado**
This pull request primarily focuses on making the application compatible with both Linux and Windows environments. The key changes involve modifying the build process, updating the GitHub workflows, and introducing new scripts for Windows. 

Here are the most important changes:

Moved Files:
* Moved `HandlerManifest.json`, `applicationhealth-shim` and `manifest.xml` to the directory `/misc/linux/`
* Moved the following files from Windows Codebase to the following directory `/misc/windows` : `HandlerManifest.json`, `disable.cmd`, `disable.ps1`, `enable.cmd`, `install.cmd`, `manifest.xml`, `resetState.cmd`, `uninstall.cmd` and `update.cmd`
Cross Platform VMWatch Changes:
* Moved previous implementation of `applyResourceGovernance` to `plugins/vmwatch/vmwatch_linux.go`
* Moved the following functions to `plugins/vmwatch/vmwatch_linux.go`: [`isSystemdAvailable`, `getSystemdVersion`, `extractVersion`, `createAndAssignCgroups`]
* Moved the Unit Test `TestExtractVersion` to `plugins/vmwatch/vmwatch_linux.go`

Cross Platform enable command handler:
* Moved the implementation of the enable function from cmdhandler_linux.go to cmdhandler.go and added Windows specific code on cmdhandler_windows.go

Windows Command Handler implementation:
* Added Windows specific implementation on cmdhandler_windows.go
* Added implementation for the enable command for Windows.
* Registry Keys: Added Pre-Function implementation for all `installHandler`, `uninstallHandler`, `enableHandler`, `updateHandler`, `disableHandler`, `resetStateHandler`. Registry Key operation is implemented on each of these functions.

Graceful Shutdown for Windows:
* Added Graceful Shutdown for Windows, by checking the state of the isEnabled registry subkey.

Updates to build process:

* [`Makefile`](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L2-R68): The build process has been split into separate commands for Linux and Windows. The `binary` command has been replaced with `binary-linux` and `binary-windows`, and similar changes have been made to other commands. The `Makefile` has also been updated to handle different file paths and manifest files for Linux and Windows. [[1]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L2-R68) [[2]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L50-R101)

Updates to GitHub workflows:

* [`.github/workflows/codeql.yml`](diffhunk://#diff-12783128521e452af0cfac94b99b8d250413c516ec71fe6d97dbea666ff7ba27L57-R57): The workflow has been updated to run on both `ubuntu-latest` and `windows-latest`. The build command has been changed from `make binary` to `make binary-linux` and `make binary-windows` depending on the OS. [[1]](diffhunk://#diff-12783128521e452af0cfac94b99b8d250413c516ec71fe6d97dbea666ff7ba27L57-R57) [[2]](diffhunk://#diff-12783128521e452af0cfac94b99b8d250413c516ec71fe6d97dbea666ff7ba27L98-R103)
* [`.github/workflows/go.yml`](diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fL15-R15): The workflow has been split into `build_linux` and `build_windows`, each with their own set of commands and environment setup. The build command has been changed from `make binary` to `make binary-linux` and `make binary-windows` depending on the job. [[1]](diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fL15-R15) [[2]](diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fL102-R102) [[3]](diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fR135-R192)

Introduction of new scripts for Windows:

* [`integration-test/env/windows/localdev.cmd`](diffhunk://#diff-41f9717ac9e048a15ef6068604ae3f9e3466606ea2898c8d4d6dd7eb241cda22R1-R19): This new script sets up the extension to be run and tested locally in Windows. It creates necessary directories, registry keys, and config files.
* [`integration-test/env/windows/localdev.settings`](diffhunk://#diff-dd75646f92864412a135a88ecbfef3991c1ac971795792123524241df9e9355eR1-R30): This new file contains extension settings for the local testing in Windows.

Other notable changes:

* [`.devcontainer/Dockerfile`](diffhunk://#diff-13bd9d7a30bf46656bc81f1ad5b908a627f9247be3f7d76df862b0578b534fc6R46-R48): Azure CLI has been added to the Dockerfile.
* [`internal/handlerenv/handlerenv.go`](diffhunk://#diff-3ccd0b34526487610cbcb81809c9cccd6ae4a738c2edd80624341773a4ff5a75L24-R31): Error handling has been improved in the `GetHandlerEnvironment` function.
* [`go.mod`](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R14): The `golang.org/x/sys` package has been updated. [[1]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R14) [[2]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L32)
* `.vscode/launch.json` and `.vscode/tasks.json`: These files have been updated to use the new `make binary-linux` command. [[1]](diffhunk://#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945R31-R49) [[2]](diffhunk://#diff-7d76d7533653c23b753fc7ce638cf64bdb5e419927d276af836d3a03fdf1745aL15-R18)This pull request primarily focuses on making the application compatible with both Linux and Windows environments. The changes include modifications to the build process, addition of new tasks for Windows, updates to the GitHub workflow, and adjustments to the grace period settings. Additionally, there are updates to the recommended extensions for the development environment.

### CODE REVIEW

File: integration-test/env/Extension/bin/VMWatch/version.txt
Comment (klugorosado) [inline integration-test/env/Extension/bin/VMWatch/version.txt:1]: Revert

File: integration-test/env/Extension/bin/VMWatch/NOTICE.txt
Comment (klugorosado) [inline integration-test/env/Extension/bin/VMWatch/NOTICE.txt:1]: Revert Probably

File: internal/manifest/extensionmanifest_windows_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	require.Equal(t, expectedManifest.IsJsonExtension, currentManifest.IsJsonExtension)
+	require.Equal(t, expectedManifest.SupportedOS, currentManifest.SupportedOS)
+	require.Equal(t, expectedManifest.CompanyName, currentManifest.CompanyName)
+}
+
+func Test_FindManifestFilePath(t *testing.T) {
+	var (
+		manifestFileName = "manifest.xml"
+		src              = "../../misc/windows"                                                 // Replace with the actual directory path
+		dst              = "/tmp/lib/waagent/Microsoft.ManagedServices-ApplicationHealthLinux/" // Replace with the actual directory path
```
Comment (dpoole73) [inline internal/manifest/extensionmanifest_windows_test.go:54]: this looks like a windows path

File: platform/cmdhandler/cmdhandler.go [hunk 2]
```diff
@@ -119,3 +128,203 @@ func printUsage(args []string) {
...
+)
+
+func enable(lg logging.Logger, h *handlerenv.HandlerEnvironment, seqNum int) (string, error) {
+	// parse the extension handler settings (not available prior to 'enable')
+	cfg, err := settings.ParseAndValidateSettings(lg, h.ConfigFolder)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get configuration")
+	}
+
+	// sendTelemetry(lg, telemetry.EventLevelInfo, telemetry.AppHealthTask, "Successfully parsed and validated settings")
```
Comment (dpoole73) [inline platform/cmdhandler/cmdhandler.go:144]: remove?

File: platform/cmdhandler/cmdhandler.go [hunk 3]
```diff
@@ -119,3 +128,203 @@ func printUsage(args []string) {
...
+	// sendTelemetry(lg, telemetry.EventLevelVerbose, telemetry.AppHealthTask, fmt.Sprintf("HandlerSettings = %s", cfg))
+
+	probe := apphealth.NewHealthProbe(lg, &cfg.AppHealthPluginSettings)
+	var (
+		intervalBetweenProbesInMs  = time.Duration(cfg.GetIntervalInSeconds()) * time.Millisecond * 1000
```
Comment (dpoole73) [inline platform/cmdhandler/cmdhandler.go:149]: is this right?  the name is confusing.  it looks like it will be a duration so it doesn't make sense for it to have units in the name.  Seems like it should just be called `intervalBetweenProbes`

File: platform/cmdhandler/cmdhandler.go [hunk 4]
```diff
@@ -119,3 +128,203 @@ func printUsage(args []string) {
...
+		numberOfProbes             = cfg.GetNumberOfProbes()
+		gracePeriodInSeconds       = time.Duration(cfg.GetGracePeriod()) * time.Second
```
Comment (dpoole73) [inline platform/cmdhandler/cmdhandler.go:151]: same deal here, shouldn't this just be gracePeriod

File: platform/cmdhandler/cmdhandler.go [hunk 5]
```diff
@@ -119,3 +128,203 @@ func printUsage(args []string) {
...
+		honorGracePeriod           = gracePeriodInSeconds > 0
+		gracePeriodStartTime       = time.Now()
+		vmWatchSettings            = cfg.GetVMWatchSettings()
+		vmWatchResult              = vmwatch.VMWatchResult{Status: vmwatch.Disabled, Error: nil}
+		vmWatchResultChannel       = make(chan vmwatch.VMWatchResult)
+		timeOfLastVMWatchLog       = time.Time{}
+	)
+
+	if !honorGracePeriod {
+		// sendTelemetry(lg, telemetry.EventLevelInfo, telemetry.AppHealthTask, "Grace period not set")
```
Comment (dpoole73) [inline platform/cmdhandler/cmdhandler.go:165]: why are all these sendTelemetry calls commented out?

File: platform/cmdhandler/cmdhandler_windows.go [hunk 6]
```diff
@@ -1,27 +1,261 @@
...
 }
 
 func (ch *WindowsCommandHandler) CommandMap() CommandMap {
 	return ch.commands
 }
+
+func installHandler(lg logging.Logger, seqNum int) error {
+	lg.Info("Installing Handler")
+	lg.Info(fmt.Sprintf(`Creating Windows Registry Key 'HKLM\%s'`, regKeyPath))
+	// Create a new registry key with all access permissions.
```
Comment (dpoole73) [inline platform/cmdhandler/cmdhandler_windows.go:133]: why all access?  does that mean anyone can read/write it?

File: plugins/vmwatch/vmwatch_linux.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	if !isSystemdAvailable() {
+		cmd = exec.Command(GetVMWatchBinaryFullPath(cmdPath), args...)
+		cmd.Env = GetVMWatchEnvironmentVariables(s.ParameterOverrides, hEnv)
+		return cmd, resourceGovernanceRequired
+	}
+
+	// if we have systemd available, we will use that to launch the process, otherwise we will launch directly and manipulate our own cgroups
+	systemdVersion := getSystemdVersion()
+
+	// since systemd-run is in different paths on different distros, we will check for systemd but not use the full path
```
Comment (dpoole73) [inline plugins/vmwatch/vmwatch_linux.go:61]: this comment is below, looks like copy patse error, remove it from here

File: plugins/vmwatch/vmwatch_test.go [hunk 8]
```diff
@@ -31,19 +31,3 @@ func TestGetMessageCorrectValue(t *testing.T) {
 	res = VMWatchResult{Status: Running}
 	require.Equal(t, "VMWatch is running", res.GetMessage())
 }
-
-func TestExtractVersion(t *testing.T) {
```
Comment (dpoole73) [inline plugins/vmwatch/vmwatch_test.go:35]: why did you remove this?

File: plugins/vmwatch/vmwatch_windows.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+		// TODO: Implement resource governance for Windows
+		resourceGovernanceRequired bool = false
+	)
+
+	cmd = exec.Command(GetVMWatchBinaryFullPath(cmdPath), args...)
+	cmd.Env = GetVMWatchEnvironmentVariables(s.ParameterOverrides, hEnv)
+	return cmd, resourceGovernanceRequired
+}
+
+// TODO: Implement resource governance for Windows
```
Comment (dpoole73) [inline plugins/vmwatch/vmwatch_windows.go:55]: i'm not sure you will really want to managed it like this since you need to create the Job Object first I think, then associate the process with the job object immediately after creating.  I guess you can refactor once you implement that

File: platform/cmdhandler/cmdhandler.go (hunk 2 above)
Comment (klugorosado) [reply platform/cmdhandler/cmdhandler.go:144]: I will Remove all the commented-out Telemetry statements on next PR, it's a pain to delete and re-add them.

File: plugins/vmwatch/vmwatch_test.go (hunk 8 above)
Comment (klugorosado) [reply plugins/vmwatch/vmwatch_test.go:35]: I moved it to vmwatch_linux_test.go since system-md is Linux specific

File: internal/manifest/extensionmanifest_windows_test.go (hunk 1 above)
Comment (klugorosado) [reply internal/manifest/extensionmanifest_windows_test.go:54]: thanks for catching this, will fix

File: platform/cmdhandler/cmdhandler.go (hunk 5 above)
Comment (klugorosado) [reply platform/cmdhandler/cmdhandler.go:165]: answered on another comment, I will fix this next PR relate to Telemetry

File: platform/cmdhandler/cmdhandler.go (hunk 3 above)
Comment (klugorosado) [reply platform/cmdhandler/cmdhandler.go:149]: this is already present on AppHealthLinux v2 but will fix

File: platform/cmdhandler/cmdhandler_windows.go (hunk 6 above)
Comment (klugorosado) [reply platform/cmdhandler/cmdhandler_windows.go:133]: No because, to access key in HKEY_LOCAL_MACHINE admin privileges are required.

File: integration-test/env/windows/HandlerEnvironment.windows.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+[
+    {
+        "version": 1,
+        "handlerEnvironment": {
```
Comment (frank-pang-msft) [inline integration-test/env/windows/HandlerEnvironment.windows.json:4]: nit: Can we name file just as HandlerEnvironment.json, since it's already in a subdirectory named windows?

File: integration-test/env/windows/localdev.settings [hunk 11]
```diff
@@ -0,0 +1,30 @@
+{
+    "runtimeSettings": [
+        {
+            "handlerSettings": {
+                "publicSettings": {
+                    "protocol": "http",
+                    "port": 8080,
+                    "requestPath": "health",
+                    "intervalInSeconds": 5,
+                    "numberOfProbes": 1,
```
Comment (frank-pang-msft) [inline integration-test/env/windows/localdev.settings:10]: nit: I know you probably just copied them over, but indentation is off

File: .github/workflows/go.yml [hunk 12]
```diff
@@ -131,4 +131,62 @@ jobs:
...
+      continue-on-error: true
+      run: test -z "$(gofmt -s -l -w $(find . -type f -name '*.go' -not -path './vendor/*') | tee /dev/stderr)"
+      working-directory: ${{ env.repo_root }}
+
+    - name: Vet
+      continue-on-error: true
+      run: test -z "$(go vet -v $(go list ./... | grep -v '/vendor/') | tee /dev/stderr)"
+      working-directory: ${{ env.repo_root }}
+
+    - uses: dominikh/staticcheck-action@v1.3.0
```
Comment (frank-pang-msft) [inline .github/workflows/go.yml:176]: this looks new, what's it for? Seems like static code check analysis?

File: internal/manifest/extensionmanifest_windows_test.go (hunk 1 above)
Comment (frank-pang-msft) [reply internal/manifest/extensionmanifest_windows_test.go:54]: still seeing linux path

File: plugins/vmwatch/vmwatch.go [hunk 13]
```diff
@@ -536,23 +356,27 @@ func GetVMWatchBinaryFullPath(processDirectory string) string {
...
+		keys []string = make([]string, 0, len(parameterOverrides)) // make sure we get the keys out in order
+	)
 
 	for k := range parameterOverrides {
 		keys = append(keys, k)
 	}
 
 	sort.Strings(keys)
-	for _, k := range keys {
+	for i, k := range keys {
```
Comment (frank-pang-msft) [inline plugins/vmwatch/vmwatch.go:374]: Is this just logging improvement? Also just wondering why we need index

File: plugins/vmwatch/vmwatch_windows.go (hunk 9 above)
Comment (frank-pang-msft) [reply plugins/vmwatch/vmwatch_windows.go:55]: +1, I think probably we can refactor later

**Review by frank-pang-msft — COMMENTED** [review]
First pass, seems most of this is just related to moving things around.

File: integration-test/env/windows/localdev.settings [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+            "handlerSettings": {
+                "publicSettings": {
+                    "protocol": "http",
+                    "port": 8080,
+                    "requestPath": "health",
+                    "intervalInSeconds": 5,
+                    "numberOfProbes": 1,
+                  "vmWatchSettings": {
+                    "enabled": true,
+                    "memoryLimitInBytes": 60000000,
```
Comment (dpoole73) [inline integration-test/env/windows/localdev.settings:13]: i think this should be 80000000

File: plugins/vmwatch/vmwatch.go [hunk 15]
```diff
@@ -527,6 +344,9 @@ func GetVMWatchConfigFullPath(processDirectory string) string {
 	return filepath.Join(processDirectory, "VMWatch", VMWatchConfigFileName)
 }
 
+// GetVMWatchBinaryFullPath returns the full path of the VMWatch binary based on the process directory.
+// It determines the binary name based on the architecture of the running process.
+// The binary name will be either VMWatchBinaryNameAmd64 or VMWatchBinaryNameArm64.s
```
Comment (dpoole73) [inline plugins/vmwatch/vmwatch.go:349]: extra `.s` on the end of string?

File: plugins/vmwatch/vmwatch.go [hunk 16]
```diff
@@ -536,23 +356,27 @@ func GetVMWatchBinaryFullPath(processDirectory string) string {
 	return filepath.Join(processDirectory, "VMWatch", binaryName)
 }
 
+// GetVMWatchEnvironmentVariables returns a list of environment variables to be attached on the VMWatch Process.
+// It takes a map of parameter overrides and a HandlerEnvironment as input.
+// It returns a slice of strings containing the environment variables.s
```
Comment (dpoole73) [inline plugins/vmwatch/vmwatch.go:361]: extra .s

File: .github/workflows/go.yml (hunk 12 above)
Comment (klugorosado) [reply .github/workflows/go.yml:176]: Yes, This is a golang linter

File: plugins/vmwatch/vmwatch.go (hunk 13 above)
Comment (klugorosado) [reply plugins/vmwatch/vmwatch.go:374]: this logs are for debugging currently, They will not be pushed the our logs files like the other logs.

Index is helpful to debug all the VMWatch Parameters.

I am thinking of deleting it all together

File: plugins/vmwatch/vmwatch_windows.go (hunk 9 above)
Comment (klugorosado) [reply plugins/vmwatch/vmwatch_windows.go:55]: Noted.

**Review by frank-pang-msft — APPROVED** [review]
I think I'm good with the changes, we'll have more confidence once you have the other PRs and we can actually test it on an actual windows machine.

**Review by dpoole73 — APPROVED** [review]
